castervoice/asynch/hud.py

```diff
@@ -0,0 +1,217 @@
+                else:
+                    gram.appendRow(rule)
+            if gram:
+                rules.appendRow(gram)
+        tree_view = QTreeView(self)
+        tree_view.setModel(rules_tree)
+        tree_view.setColumnWidth(0, RulesWindow._WIDTH / 2)
+        layout = QVBoxLayout()
+        layout.addWidget(tree_view)
+        self.setLayout(layout)
+
+
+class HUDWindow(QMainWindow):
+
+    _WIDTH = 300
+    _HEIGHT = 200
+    _MARGIN = 30
+
+    def __init__(self, server):
+        QMainWindow.__init__(self, flags=(PySide2.QtCore.Qt.WindowStaysOnTopHint))
+        x = dragonfly.monitors[0].rectangle.dx - (HUDWindow._WIDTH + HUDWindow._MARGIN)
+        y = HUDWindow._MARGIN
+        dx = HUDWindow._WIDTH
+        dy = HUDWindow._HEIGHT
+        self.server = server
+        self.setup_xmlrpc_server()
+        self.setGeometry(x, y, dx, dy)
+        self.setWindowTitle("Caster HUD")
+        self.output = QTextEdit()
+        self.output.setReadOnly(True)
+        self.setCentralWidget(self.output)
+        self.rules_window = None
+        self.commands_count = 0
+
+    def event(self, event):
+        if event.type() == SHOW_HUD_EVENT:
+            self.show()
+            return True
+        if event.type() == HIDE_HUD_EVENT:
+            self.hide()
+            return True
+        if event.type() == SHOW_RULES_EVENT:
+            self.rules_window = RulesWindow(event.text)
+            self.rules_window.show()
+            return True
+        if event.type() == HIDE_RULES_EVENT and self.rules_window:
+            self.rules_window.close()
+            self.rules_window = None
+            return True
+        if event.type() == SEND_COMMAND_EVENT:
+            escaped_text = html.escape(event.text)
+            if escaped_text.startswith('$'):
+                formatted_text = '<font color="blue">&lt;</font><b>{}</b>'.format(escaped_text[1:])
+                if self.commands_count == 0:
+                    self.output.setHtml(formatted_text)
+                else:
+                    # self.output.append('<br>')
+                    self.output.append(formatted_text)
+                cursor = self.output.textCursor()
+                cursor.movePosition(PySide2.QtGui.QTextCursor.End)
+                self.output.setTextCursor(cursor)
+                self.output.ensureCursorVisible()
+                self.commands_count += 1
+                if self.commands_count == 20:
+                    self.commands_count = 0
+                return True
+            elif escaped_text.startswith('#'):
+                formatted_text = '<font color="red">&gt;</font>{}'.format(escaped_text[1:])
+            else:
+                formatted_text = escaped_text
+            self.output.append(formatted_text)
+            self.output.ensureCursorVisible()
+            return True
+        return QMainWindow.event(self, event)
+
+    def closeEvent(self, event):
+        event.accept()
+
+    def setup_xmlrpc_server(self):
```

Comment:
Apologies for the drive-by comment, but might this need a security token type of thing here, like what dragonfly & aenea use? https://github.com/dictation-toolbox/dragonfly/pull/61/files#diff-6219d462fba21473f0a2404097943f75R29

The danger case would be something like: user is tricked into browsing to evil website, evil website asks web browser to make xmlhttprequests to http://localhost:1338, evil website can now ask this server to execute any registered function. Obviously from a practical perspective this is only going to be a concern if a caster user is actively being targeted (and the blast radius in this server seems small right now - basically it'd be annoying to have the rules UI show up or have caster quit unexpectedly), but mentioning it here anyway.

Reply:
Thanks. I will consider adding the security token. Probably, not in this PR, however.

Reply:
It adding the security token difficult? I wouldn't want any extra vulnerabilities introduced to master.

Reply:
It's actually a vulnerability that exists already with the current implementation.

Reply:
The current master you mean? Can we fix that?

Reply:
@kendonB That's correct and in the current master. Basically someone would have to set up a malicious website and a caster user would have to go to that website. In that case they then could control RPC commands that are defined. Mostly emulating mouse now this is none that I know of control keyboard input.

To protect against this a secret would be randomly generated and be used for authentication for the client/server to protect against such targeting.

dmakarov mentioned he would implement it in another pull request.

Reply:
Yes, sorry, I should have mentioned that this is an existing problem so it's not meaningfully worse than before (just a couple new commands available to an attacker - but none of them seem to allow arbitrary code execution).

Reply:
@caspark would you be able to open an issue on this (for the current vulnerabilities)?

Reply:
Thanks for the reminder @caspark
Documented in https://github.com/dictation-toolbox/Caster/issues/809

Reply:
> 
> 
> Thanks. I will consider adding the security token. Probably, not in this PR, however.

Is this something you are willing to take on?